amivapi/tests/studydocs/test_rating.py

```diff
@@ -0,0 +1,229 @@
+# -*- coding: utf-8 -*-
+#
+# license: AGPLv3, see LICENSE for details. In addition we strongly encourage
+#          you to buy us beer if we meet and you like the software.
+"""Tests for studydocuments rating."""
+
+from amivapi.studydocs.rating import compute_rating
+from amivapi.tests.utils import WebTest, WebTestNoAuth
+
+
+class StudydocsRatingTest(WebTestNoAuth):
+    """Test rating computation."""
+
+    def test_no_rating(self):
+        """Without any votes, the `rating` is in the document, and is None."""
+        doc = self.new_object('studydocuments')
+
+        response = self.api.get('/studydocuments/' + str(doc['_id']),
+                                status_code=200)
+
+        self.assertIsNone(response.json['rating'])
+
+    def test_bound(self):
+        """Without any votes, the rating is None."""
+        for upvotes, downvotes, rating in [
+            (1, 0, 0.38),
+            (5, 0, 0.75),
+            (10, 0, 0.86),
+            (1, 1, 0.16),
+            (5, 2, 0.47),
+            (5, 5, 0.31),
+            (50, 50, 0.44),
+            (0, 1, 0.0),
+            (0, 5, 0.0),
+            (0, 10, 0.0),
+        ]:
+            bound = compute_rating(upvotes, downvotes)
+            self.assertAlmostEqual(rating, bound, 2)
+
+    def test_rating(self):
+        """Test that the rating is correctly attached to the studydoc."""
+        doc = self.new_object('studydocuments')
+        lookup = {'studydocument': str(doc['_id'])}
+
+        response = self.api.get('/studydocuments/' + str(doc['_id']),
+                                status_code=200)
+        # No ratings yet
+        self.assertIsNone(response.json['rating'])
+
+        self.load_fixture({
+            'users': [{}],
+            'studydocumentratings': [{'rating': 'up', **lookup}],
+        })
+
+        response = self.api.get('/studydocuments/' + str(doc['_id']),
+                                status_code=200)
+        self.assertAlmostEqual(response.json['rating'], 0.38, 2)
+
+    def test_sort_by_rating(self):
+        """Ensure that sorting by rating works."""
+        ids = [24 * '0', 24 * '1', 24 * '2']
+        self.load_fixture({
+            'users': [{}],
+            'studydocuments': [
+                {'_id': ids[0]}, {'_id': ids[1]}, {'_id': ids[2]}
+            ],
+            'studydocumentratings': [
+                # Rate such that:
+                # 0: No rating, 1: Low rating, 2: High rating
```

Comment:
This is a slightly weird ordering... But that is just what the math works out to. We could in theory give unrated things a slightly positive epsilon rating to put them above things only voted negative.